ws/adapter.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) Mainflux
+// SPDX-Licence-Identifier: Apache-2.0
+
+// Package ws contains the domain concept definitions needed to support
+// Mainflux ws adapter service functionality
+
+package ws
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/mainflux/mainflux"
+	"github.com/mainflux/mainflux/pkg/errors"
+	"github.com/mainflux/mainflux/pkg/messaging"
+)
+
+var (
+	// ErrFailedMessagePublish indicates that message publishing failed.
+	ErrFailedMessagePublish = errors.New("failed to publish message")
+
+	// ErrFailedSubscription indicates that client couldn't subscriber to specified channel
+	ErrFailedSubscription = errors.New("failed to subscribe to a channel")
+
+	// ErrFailedConnection indicates that service couldn't connect to message broker.
+	ErrFailedConnection = errors.New("failed to connect to message broker")
+
+	// ErrInvalidConnection indicates that client couldn't subscribe to message broker
+	ErrInvalidConnection = errors.New("nats: invalid connection")
+
+	// ErrAlreadySubscribed indicates that client couldn't subscribe, as it was already subscribed
+	ErrAlreadySubscribed = errors.New("already subscribed to topic")
+
+	// ErrUnauthorizedAccesss indicates that client provided missing or invalid credentials
+	ErrUnauthorizedAccess = errors.New("missing or invalid credentials provided")
+
+	// ErrEmptyTopic and ErrEmptyID indicate absence of channelID or thingKey in the request
+	ErrEmptyTopic = errors.New("empty topic")
+	ErrEmptyID    = errors.New("empty id")
+)
+
+// Service specifies web socket service API.
+type Service interface {
+	// Publish Message
+	Publish(ctx context.Context, thingKey string, msg messaging.Message) error
+
+	// Subscribes to a channel with specified id.
+	Subscribe(ctx context.Context, thingKey, chanID, subtopic string, client messaging.MessageHandler) error
+
+	// Unsubscribe method is used to stop observing resource.
+	Unsubscribe(ctx context.Context, thingKey, chanID, subtopic string) error
+
+	Authorize(context.Context, string, string) (*mainflux.ThingID, error)
```

Comment:
I don't like the idea of exposing the `Authorize` method. The adapter should take care of it silently, without exposing it for general use.

Reply:
This was because I couldn't use it without exposing. (After the change to start using MessageHandler interface in `Subscribe()` function above this).
I don't currently know how to fix this problem, I'll try this tomorrow.